src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -0,0 +1,329 @@
+/// object table entries up to but not including this notification.
+bool UpdateObjectLocations(bool is_added,
+                           const std::vector<ObjectTableData> &location_updates,
+                           std::shared_ptr<gcs::GcsClient> gcs_client,
+                           std::unordered_set<ClientID> *node_ids) {
+  // location_updates contains the updates of locations of the object.
+  // with GcsChangeMode, we can determine whether the update mode is
+  // addition or deletion.
+  bool isUpdated = false;
+  for (const auto &object_table_data : location_updates) {
+    ClientID node_id = ClientID::FromBinary(object_table_data.manager());
+    if (is_added && 0 == node_ids->count(node_id)) {
+      node_ids->insert(node_id);
+      isUpdated = true;
+    } else if (!is_added && 1 == node_ids->count(node_id)) {
+      node_ids->erase(node_id);
+      isUpdated = true;
+    }
+  }
+  // Filter out the removed clients from the object locations.
+  for (auto it = node_ids->begin(); it != node_ids->end();) {
+    if (gcs_client->Nodes().IsRemoved(*it)) {
+      it = node_ids->erase(it);
+    } else {
+      it++;
+    }
+  }
+
+  return isUpdated;
+}
+
+}  // namespace
+
+ray::Status OwnershipBasedObjectDirectory::ReportObjectAdded(
+    const ObjectID &object_id, const ClientID &client_id,
+    const object_manager::protocol::ObjectInfoT &object_info) {
+  WorkerID worker_id = WorkerID::FromBinary(object_info.owner_worker_id);
+  auto it = worker_rpc_clients_.find(worker_id);
+  if (it == worker_rpc_clients_.end()) {
+    rpc::Address owner_address;
+    owner_address.set_raylet_id(object_info.owner_raylet_id);
+    owner_address.set_ip_address(object_info.owner_ip_address);
+    owner_address.set_port(object_info.owner_port);
+    owner_address.set_worker_id(object_info.owner_worker_id);
+    auto client = std::unique_ptr<rpc::CoreWorkerClient>(
+        new rpc::CoreWorkerClient(owner_address, client_call_manager_));
+    it = worker_rpc_clients_
+             .emplace(worker_id,
+                      std::make_pair<std::unique_ptr<rpc::CoreWorkerClient>, size_t>(
+                          std::move(client), 0))
+             .first;
+  }
+
+  rpc::AddObjectLocationOwnerRequest request;
+  request.set_intended_worker_id(object_info.owner_worker_id);
+  request.set_object_id(object_id.Binary());
+  request.set_client_id(client_id.Binary());
+
+  worker_rpc_clients_[worker_id].second++;
+  RAY_CHECK_OK(it->second.first->AddObjectLocationOwner(
+      request, [this, worker_id, object_id](
+                   Status status, const rpc::AddObjectLocationOwnerReply &reply) {
+        if (!status.ok()) {
+          RAY_LOG(ERROR) << "Worker " << worker_id << " failed to add the location for "
+                         << object_id;
+        }
+        // Remove the cached worker client if there are no more pending requests.
+        if (--worker_rpc_clients_[worker_id].second == 0) {
```

Comment:
Let's just keep the clients forever for now. This will be fixed with @barakmich 's connection pooling in the future.